### ISSUE TITLE
(🐞) `type.__call__` not checked properly

### ISSUE THREAD

**Issue by KotlinIsland** [python/mypy#17316]
```py
class A:
    def __init__(self, x: int):
        print(x)

A.__call__("a")  # no error
A("a")  # Argument 1 to "A" has incompatible type "str"; expected "int"  [arg-type]
```

**Comment by erictraut** [comment]
The `type` class in the `builtins.pyi` stub file is defined as:

```python
class type:
     ...
    def __init__(self, name: str, bases: tuple[type, ...], dict: dict[str, Any], /, **kwds: Any) -> None: ...
     ...
    def __call__(self, *args: Any, **kwds: Any) -> Any: ...
```

Based on this definition, mypy is working correctly. Not surprisingly, its behavior matches the other major type checkers here.

If you think that typeshed definition is incorrect or could be improved, you could file a bug report or PR in the typeshed project.

**Comment by KotlinIsland** [comment]
@traut, if the type checker were just following the stubs, then I would expect `A("a")` to behave identically to `A.__call__("a")` and not show an error. The truth is that currently type checkers are special casing `type.__call__`, but only the operator form, and not the attribute form.

**Comment by erictraut** [comment]
> The truth is that currently type checkers are special casing type.__call__, but only the operator form, and not the attribute form.

Yes, that's conformant with the recently accepted [typing spec chapter on constructor call validation](https://typing.readthedocs.io/en/latest/spec/constructors.html#constructor-calls). Mypy is doing the right thing here according to the typing spec, and so are the other type checkers.

If you would like to suggest a change to the typing spec, the [typing forum](https://discuss.python.org/c/typing/32) would be a good place to have that discussion.

**Comment by KotlinIsland** [comment]
> Type checkers should mirror this runtime behavior when analyzing a constructor call.

Type checkers are not mirroring this runtime behavior correctly in this case. 

That spec should be updated to be more explicit about how to mirror the runtime behavior (to match this case), the type-checkers should be updated to match the runtime behaviour, and typeshed could be updated to have the correct type.
